Add reader for FY-3E / MERSI-LL

simonrp84 opened the pull request:
This PR adds a reader for the MERSI-LowLight sensor aboard the FengYun-3E satellite. It's currently a draft as there's no additional tests but I aim to add those soon.

 - [x] Tests added

simonrp84 commented:
pre-commit.ci autofix

TAlonglong commented:
Could this be used to have the moon_phase:

```python
>>> import ephem
>>> import datetime
>>> now = datetime.datetime.now()
>>> now
datetime.datetime(2023, 2, 28, 18, 3, 47, 660535)
>>> m = ephem.Moon(now)
>>> m.moon_phase
0.6374500774243032
```

TAlonglong commented:
I get this when reading 250meter resolution data. I will have a look
```
/home/trygveas/miniconda3/envs/python/lib/python3.9/site-packages/paramiko/transport.py:219: CryptographyDeprecationWarning: Blowfish has been deprecated
  "class": algorithms.Blowfish,
Failed to load DataID(name='6', wavelength=WavelengthRange(min=10.3, central=10.8, max=11.3, unit='µm'), resolution=250, calibration=<calibration.brightness_temperature>, modifiers=()) from <MERSI2L1B: '/home/trygveas/testdata/fy3/FY3E_X_2023_01_11_15_12_A_G_MERSI_0250M_L1B.HDF'>
Traceback (most recent call last):
  File "/home/trygveas/pytroll/satpy/satpy/readers/yaml_reader.py", line 699, in _load_dataset
    projectable = fh.get_dataset(dsid, ds_info)
  File "/home/trygveas/pytroll/satpy/satpy/readers/mersi2_l1b.py", line 124, in get_dataset
    data = self._get_bt_dataset(data, calibration_index, wave_number)
  File "/home/trygveas/pytroll/satpy/satpy/readers/mersi2_l1b.py", line 195, in _get_bt_dataset
    coeffs = self['/attr/TBB_Trans_Coefficient']
  File "/home/trygveas/pytroll/satpy/satpy/readers/hdf5_utils.py", line 96, in __getitem__
    val = self.file_content[key]
KeyError: '/attr/TBB_Trans_Coefficient'
```

TAlonglong commented:
Hm If I do something like this:
```python
        elif self.sensor_name == 'mersi-ll':
            # MERSI-LL stores these coefficients differently
            try:
                coeffs = self['/attr/TBB_Trans_Coefficient']
                corr_coeff_a = coeffs[calibration_index]
                corr_coeff_b = coeffs[calibration_index + N_TOT_IR_CHANS_LL]
            except KeyError:
                return data
```

I get the data back. But I don't know if there is any science in the data after this.

djhoese commented:
Just curious @simonrp84 @TAlonglong what's the status of this PR? My bosses have started asking if I can add this reader soon.

simonrp84 commented:
@djhoese To be honest this slipped my mind. I added the reader 'for fun' rather than for a specific work project so can't guarantee I'll finish it fast. Am away until 11th April but will look then and give you an estimate of when I might be able to have it done.

simonrp84 commented:
Looks like recent FY-3E data no longer contains the calibration coefficients (geez, I love consistency!)
So, @TAlonglong's "fix" with the `try` and `except` does the job it seems.

djhoese commented on satpy/tests/reader_tests/test_mersi2_l1b.py:
```diff
@@ -513,31 +531,195 @@ def test_fy3d_250_resolutions(self):
+        assert res['25'].attrs['units'] == 'K'
+
+
+class TestFY3EMERSI2L1B(MERSI2L1BTester):
```
Thoughts on adding `LL` to the name? Too hard to read? I just wanted to make it extra clear this is for a similar yet completely different instrument to the other MERSI2 instrument.

```suggestion
class TestFY3EMERSI2LLL1B(MERSI2L1BTester):
```

simonrp84 replied:
This is a mistake, it should be `TestFY3EMERSILLL1B`. Which is also fairly incomprehensible. Can change that if you have a better suggestion.

djhoese replied:
Ah with the clarification that the instrument is MERSI-LL and not MERSI-2-LL, maybe just remove the `2` from the existing name?

djhoese commented on satpy/etc/readers/mersi_ll_l1b.yaml:
```diff
@@ -0,0 +1,251 @@
+reader:
+  description: FY-3D Medium Resolution Spectral Imager 2 (MERSI-2) L1B Reader
+  name: mersi_ll_l1b
+  sensors: [mersi-ll]
```
Should this be `mersi-ll` or `mersi-2-ll` or `mersi2-ll`? The `2` is important isn't it?

simonrp84 replied:
No, they are totally separate instruments and I think the names are correct. FY-3D and FY-3F fly the `MERSI-2` instrument while FY-3E flies `MERSI-LL`

That said, your comment has made me realise I forgot to change the description. It's the copy-pasted one from the mersi-2 yaml. Will fix that later today!

djhoese replied:
Ah gotcha. Sounds good then.

djhoese reviewed (commented):
I'm surprised the python code could be reused so much. So I'm glad that you were able to reuse so much code, but also sad because it makes decisions on how to split up tests and YAML files a little confusing. I think you've got a pretty good balance here.

I left a couple inline comments. The biggest one is about what the sensor name should be. I found it odd that the other instrument is `mersi-2` and this one is `mersi-ll`.

simonrp84 commented:
Ok, I updated the names - hopefully the new one is better suited!

djhoese reviewed (changes requested):
Should probably rename `test_mersi2_l1b.py` to `test_mersi_l1b.py` to match the reader python module, right?

And is this ready to be marked as "Ready for Review"?

simonrp84 commented:
> Should probably rename `test_mersi2_l1b.py` to `test_mersi_l1b.py` to match the reader python module, right?
> 
> And is this ready to be marked as "Ready for Review"?

Ooops, yes to both. Updated now.

djhoese commented on satpy/readers/mersi_l1b.py:
```diff
@@ -179,12 +181,24 @@ def _get_bt_dataset(self, data, calibration_index, wave_number):
+
+        # Some BT bands seem to have 0 in the first 10 columns
+        # and it is an invalid measurement, so let's mask
+        data = data.where(data != 0)
```
Any idea if this is for all MERSI files (mersi 2 and LL)? Is it something that might be fixed? Could we put an `if` statement around this to only apply it for LL if that is the case?

simonrp84 replied:
This was in the original MERSI-2 reader, and I have seen it occur with MERSI-2 files.

Haven't seen it happen with MERSI-LL, but then I've only examined a relatively small number of files so I can't definitively say that it won't happen with this instrument too!

djhoese replied:
Ok, let's leave it here then.

djhoese replied:
Ok one issue I see with this is that the previous version of this was applied *after* the correction coefficients were applied. I guess since the data is already supposed to be BTs (and not radiances) then this is probably still OK, but I was wondering if it might be "more" accurate to have it after the coefficients are applied?

simonrp84 replied:
Changed it back for consistency.

djhoese commented on satpy/etc/composites/mersi-ll.yaml:
```diff
@@ -0,0 +1,91 @@
+    standard_name: equalized_radiance
+    units: "1"
+
+  # These are commented out currently due to lack of Moon phase information
+  #dynamic_dnb:
+  #  compositor: !!python/name:satpy.composites.viirs.ERFDNB
+  #  prerequisites:
+  #  - name: '1'
+  #  - name: 'solar_zenith_angle'
+  #  - name: 'moon_zenith_angle'
+  #  standard_name: equalized_radiance
+  #  units: "1"
+
+  #hncc_dnb:
+  #  compositor: !!python/name:satpy.composites.viirs.NCCZinke
+  #  prerequisites:
+  #  - name: '1'
+  #  - name: 'solar_zenith_angle'
+  #  - name: 'moon_zenith_angle'
+  #  standard_name: ncc_radiance
+  #  units: "1"
```
Is the moon angle information still not available? I see some moon "stuff" in the reader YAML.

djhoese commented on satpy/etc/readers/mersi_ll_l1b.yaml:
```diff
@@ -0,0 +1,251 @@
+      - '{filename_mda}_MERSI_GEOQK_L1B.{ext:3s}'
+
+
+# NOTE: OSCAR website currently has bands in wavelength order
+#       https://www.wmo-sat.info/oscar/instruments/view/279
+#       The order below is by the wavelength in the input files
+#       The slides at the below link have band 5 and 19 swapped:
+#       http://www.wmo.int/pages/prog/sat/meetings/documents/IPET-SUP-4_Doc_05-04_FY-3D-ppt.pdf
```
How much do we know about what might change with the band naming/ordering? Should we wait a bit longer to merge this? Should we use different band names than just the numbers?

simonrp84 replied:
This was from the original MERSI-2 reader, it's not something I added.

I should delete this comment from the YAML file, as it only applies to MERSI-2.

djhoese replied:
Yeah I guess it would make sense to get rid of it for the LL reader. Thanks.

simonrp84 commented on satpy/readers/mersi_l1b.py:
```diff
@@ -179,12 +181,24 @@ def _get_bt_dataset(self, data, calibration_index, wave_number):
+        elif self.sensor_name == 'mersi-ll':
+            # MERSI-LL stores these coefficients differently
+            try:
+                coeffs = self['/attr/TBB_Trans_Coefficient']
```
Older data (both pre-operational and operational) have these coefficients. More recent data (some time in 2022, can't find exact date) do not have these coefficients. So unfortunately it's not just test data.

TAlonglong commented:
Sorry to interrupt here but I see 
```
h5dump -H /home/trygveas/testdata/fy3/FY3E_X_2023_01_11_15_12_A_G_MERSI_GEO1K_L1B.HDF | grep Moon
      DATASET "MoonAzimuth" {
      DATASET "MoonZenith" {
```

simonrp84 commented:
Interesting, how are you retrieving those files? Are they from direct broadcast or the CMA datacentre? All of mine are from CMA - so if yours are DB then maybe there's an inconsistency...

TAlonglong commented:
Yeah its from Direct Broadcast processed with the CMA software FY3E package.

djhoese commented on satpy/readers/mersi_l1b.py:
```diff
@@ -107,12 +109,12 @@ def get_dataset(self, dataset_id, ds_info):
+            coeffs = self._get_coefficients(ds_info['calibration_key'],
+                                            ds_info['calibration_index'])
+            data = coeffs[0] + coeffs[1] * data + coeffs[2] * data ** 2
             # some bands have 0 counts for the first N columns and
             # seem to be invalid data points
             data = data.where(data != 0)
```
This line of code used to be only in the BT calibration function...right? Now it is in the reflectance section?

simonrp84 replied:
Geez, yes it did. Fixed.

djhoese reviewed (approved):
I think I'm OK with this being merged pending whatever you and @TAlonglong figure out between DB and CMA versions of the files. Regardless we could merge it and then let people play with it and work on the DNB-like composites/normalizations and make updates as the file format changes. What do you two think?

TAlonglong commented:
I'm fine with that.

simonrp84 commented:
i've now updated the HNCC composite to cope with missing Moon phase data, it should calculate it internally. Tests are also updated to test as much of this as I can figure out how to do :)

simonrp84 commented:
Codefactor isn't happy, but those appear to be existing problems rather than something new.

simonrp84 commented on satpy/composites/viirs.py:
```diff
@@ -879,8 +879,18 @@ class NCCZinke(CompositeBase):
+            moon_illum_fraction = ephem.Moon(datasets[0].attrs['start_time']).moon_phase
+        elif len(datasets) == 4:
+            # convert to decimal instead of %
+            moon_illum_fraction = da.mean(datasets[3].data) * 0.01
```
Thanks @djhoese. Have done both of those now - tests pass locally so as long as they're good in the CI then hopefully this is now ready to go.

mraspaud commented on satpy/composites/viirs.py:
```diff
@@ -865,6 +865,24 @@ def _linear_normalization_from_0to1(
 
+def _check_moon_phase(datasets):
+    """Check if we have Moon phase as an input dataset and, if not, calculate it."""
+    if len(datasets) < 3 or len(datasets) > 4:
+        raise ValueError("Expected either 3 or 4 datasets, got %d" % (len(datasets),))
```
I think it's a bit confusing that the argument checks is done in this function. I would rather have the calling function call eg `_check_three_or_four_arguments(datasets)`that just does the argument checking and then `moon_illum_fraction = get_moom_illum_fraction(dataset[3:4])`
(`dataset[3:4]`gives `[]`if there are only 3 elements in the list)

djhoese replied:
This should be done now. Let me know what you think.

djhoese commented:
Yeah that refactor of using fixtures for these reader tests will take a long time. They are all based on VIIRS and MODIS (I think) tests I wrote a long time ago before we were using pytest. They really only work for file handlers using the "utility file handlers" for reading NetCDF and HDF files, but yeah a fixture would likely be much cleaner.
